lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,351 @@
+    }
+
+    private val payloadCallback = object : PayloadCallback() {
+        override fun onPayloadReceived(endpointId: String, payload: Payload) {
+            when (payload.type) {
+                Payload.Type.BYTES -> handleBytesPayload(endpointId, payload)
+                else -> log("Received unsupported payload type from: $endpointId")
+            }
+        }
+
+        override fun onPayloadTransferUpdate(endpointId: String, update: PayloadTransferUpdate) {
+        }
+    }
+
+    private fun updateEndpointIpAddress(endpointId: String, ipAddress: String) {
+        _endpointStatusFlow.update { currentMap ->
+            currentMap.toMutableMap().apply {
+                this[endpointId]?.ipAddress = ipAddress
+            }
+        }
+    }
+
+    private fun updateEndpointStatus(endpointId: String, status: EndpointStatus) {
+        _endpointStatusFlow.update { currentMap ->
+            currentMap.toMutableMap().apply {
+                this[endpointId]?.status = status
+            }
+        }
+    }
+
+    private fun addEndpoint(endpointId: String) {
+        _endpointStatusFlow.update { currentMap ->
+            currentMap.toMutableMap().apply {
+                this[endpointId] = EndpointInfo(endpointId)
+            }
+        }
+    }
+
+    private fun removeEndpoint(endpointId: String) {
+        _endpointStatusFlow.update { currentMap ->
+            currentMap.toMutableMap().apply {
+                remove(endpointId)
+            }
+        }
+    }
+
+    private fun checkAndInitiateConnection(endpointId: String) {
+        val endpointInfo = _endpointStatusFlow.value[endpointId]
+        if (endpointInfo != null && endpointInfo.status == EndpointStatus.DISCONNECTED) {
+            requestConnection(endpointId)
+        }
+    }
+
+    private fun requestConnection(endpointId: String) {
+        connectionsClient.requestConnection(name, endpointId, connectionLifecycleCallback)
+            .addOnSuccessListener {
+                log("Connection request sent to endpoint: $endpointId")
+                updateEndpointStatus(endpointId, EndpointStatus.CONNECTING)
+                _endpointStatusFlow.update { currentMap ->
+                    currentMap.toMutableMap().apply {
+                        this[endpointId]?.isOutgoing = true
+                    }
+                }
+            }
+            .addOnFailureListener { e ->
+                log("Failed to request connection to endpoint: $endpointId", e)
+                updateEndpointStatus(endpointId, EndpointStatus.DISCONNECTED)
+            }
+    }
+
+    private fun log(message: String, exception: Exception? = null) {
+        val prefix = "[NearbyVirtualNetwork:$name] "
+        if (exception != null) {
+            logger(Log.ERROR, "$prefix$message", exception)
+        } else {
+            logger(Log.DEBUG, "$prefix$message")
+        }
+    }
+
+    private fun observeEndpointStatusFlow() {
+        CoroutineScope(Dispatchers.Default).launch {
```

Comment:
You need to create a scope (as a member val) which should then be canceled by calling close.